File: windows/Hinge.Core/LruRingCache.cs
namespace Hinge.Core;

/// <summary>
/// Thread-safe LRU ring cache for anti-looping deduplication (default capacity: 100).
/// </summary>
public class LruRingCache
{
    private readonly int _capacity;
    private readonly object _lock = new();
    private readonly LinkedList<string> _order = new();
    private readonly HashSet<string> _lookup = new();

    public int Capacity => _capacity;
    public int Count
    {
        get
        {
            lock (_lock)
            {
                return _lookup.Count;
            }
        }
    }

    public LruRingCache(int capacity = 100)
    {
        if (capacity <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(capacity), "Capacity must be positive.");
        }
        _capacity = capacity;
    }

    /// <summary>
    /// Checks whether the cache contains the specified identifier.
    /// </summary>
    public bool Contains(string id)
    {
        if (string.IsNullOrEmpty(id)) return false;
        lock (_lock)
        {
            return _lookup.Contains(id);
        }
    }

    /// <summary>
    /// Adds an identifier to the cache. If already present, refreshes its LRU position.
    /// If capacity is exceeded, evicts the oldest item.
    /// Returns true if newly added; false if already present.
    /// </summary>
    public bool Add(string id)
    {
        if (string.IsNullOrEmpty(id)) return false;

        lock (_lock)
        {
            if (_lookup.Contains(id))
            {
                _order.Remove(id);
                _order.AddLast(id);
                return false;
            }

            if (_lookup.Count >= _capacity)
            {
                var oldest = _order.First;
                if (oldest != null)
                {
                    _lookup.Remove(oldest.Value);
                    _order.RemoveFirst();
                }
            }

            _lookup.Add(id);
            _order.AddLast(id);
            return true;
        }
    }

    /// <summary>
    /// Clears the cache.
    /// </summary>
    public void Clear()
    {
        lock (_lock)
        {
            _lookup.Clear();
            _order.Clear();
        }
    }
}

File: android/lib/core/lru_ring_cache.dart
import 'dart:collection';

/// LRU Ring Cache for anti-looping deduplication (default capacity: 100).
class LruRingCache {
  final int capacity;
  final LinkedHashSet<String> _set = LinkedHashSet<String>();

  LruRingCache([this.capacity = 100]) {
    if (capacity <= 0) {
      throw ArgumentError.value(
        capacity,
        'capacity',
        'Capacity must be positive.',
      );
    }
  }

  int get count => _set.length;

  bool contains(String id) {
    if (id.isEmpty) return false;
    return _set.contains(id);
  }

  /// Adds an identifier to the cache.
  /// If already present, moves to the most recent end and returns false.
  /// If newly added and capacity is exceeded, evicts the oldest item.
  /// Returns true if newly added, false if already present.
  bool add(String id) {
    if (id.isEmpty) return false;

    if (_set.contains(id)) {
      _set.remove(id);
      _set.add(id);
      return false;
    }

    if (_set.length >= capacity) {
      final oldest = _set.first;
      _set.remove(oldest);
    }

    _set.add(id);
    return true;
  }

  void clear() {
    _set.clear();
  }
}
